core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -126,7 +126,7 @@ public RESTSessionCatalog() {
     this(config -> HTTPClient.builder(config).uri(config.get(CatalogProperties.URI)).build());
   }
 
-  RESTSessionCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
+  public RESTSessionCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
```

Comment:
In general, I'm okay with this approach. I think setting these in the constructor is probably the right way to go if we need to do more of this. We should also remove the new setter method if we're opening this up and add the builder here.

I'm also curious why this is needed, since I think that catalogs probably will have their own client and the set of catalogs doesn't really change in Trino. What do you intend to do differently when you can use this?

Reply:
The Trino FileIO wrapper is created per Trino query, since it has context about the user (for example, HDFS with Kerberos impersonation). Thus, we need to create the Iceberg catalog per query as well.

The mismatch is the scope of the services. FileIO has query scope, but the HTTP client has singleton scope.

We need to determine the appropriate scope for everything inside the REST catalog and make it injectable.

Reply:
Okay, I think we settled this offline. We can move forward with something better for REST since we already handle sessions for REST and don't need a catalog instance per user.